### ISSUE TITLE
Containers are created but not used

### ISSUE THREAD

**Issue by asadpiz** [jenkinsci/docker-plugin#738]
- docker-plugin version 1.1.6
 - Jenkins ver. 2.176.1
 - Docker version 18.06.1-ce

I'm trying to run a simple git clone command inside a container (jenkins-ssh), which is on another host (slave host):
pipeline {
    agent {label 'slave1'} 
    stages {
        stage('GetCode') { 
            steps {
                sh 'git clone myrepo.git'
            }
....

The problem is that although container are created successfully, docker ps + docker logs. The jenkins UI shows the message that **"Queued All nodes of label ‘slave1’ are offline"**. 

Jenkins logs are showing the following:

```
INFO: Trying to run container for node slave1-0007qgo2yrd6u from image: jenkins/ssh-slave
Jul 03, 2019 1:39:18 PM com.nirima.jenkins.plugins.docker.DockerTemplate doProvisionNode
INFO: Started container ID 42bac5ecbe3e6b81bfc25822d62a542628745c602198e80ef7fdbe68bfdc95b0 for node slave1-0007qgo2yrd6u from image: jenkins/ssh-slave
Jul 03, 2019 1:39:19 PM com.nirima.jenkins.plugins.docker.utils.PortUtils$ConnectionCheckSSH execute
INFO: SSH port is open on 192.168.1.18:32846
Jul 03, 2019 1:39:27 PM hudson.slaves.NodeProvisioner$2 run
INFO: Image of jenkins/ssh-slave provisioning successfully completed. We have now 31 computer(s)
Jul 03, 2019 1:39:27 PM com.nirima.jenkins.plugins.docker.DockerCloud provision
INFO: Asked to provision 1 slave(s) for: slave1
Jul 03, 2019 1:39:27 PM com.nirima.jenkins.plugins.docker.DockerCloud canAddProvisionedSlave
INFO: Provisioning 'jenkins/ssh-slave' on 'docker'; Total containers: 30 (of 100)
Jul 03, 2019 1:39:27 PM com.nirima.jenkins.plugins.docker.DockerCloud provision
INFO: Will provision 'jenkins/ssh-slave', for label: 'slave1', in cloud: 'docker'
Jul 03, 2019 1:39:27 PM hudson.slaves.NodeProvisioner$StandardStrategyImpl apply
INFO: Started provisioning Image of jenkins/ssh-slave from docker with 1 executors. Remaining excess workload: 0
Jul 03, 2019 1:39:27 PM com.nirima.jenkins.plugins.docker.DockerTemplate pullImage
INFO: Pulling image 'jenkins/ssh-slave:latest'. This may take awhile...
Jul 03, 2019 1:39:28 PM com.nirima.jenkins.plugins.docker.DockerTemplate pullImage
INFO: Finished pulling image 'jenkins/ssh-slave:latest', took 1115 ms
Jul 03, 2019 1:39:28 PM com.nirima.jenkins.plugins.docker.DockerTemplate doProvisionNode
INFO: Trying to run container for jenkins/ssh-slave
Jul 03, 2019 1:39:28 PM com.nirima.jenkins.plugins.docker.DockerTemplate doProvisionNode
INFO: Trying to run container for node slave1-0007qgsnyzliy from image: jenkins/ssh-slave
Jul 03, 2019 1:39:28 PM com.nirima.jenkins.plugins.docker.DockerTemplate doProvisionNode
INFO: Started container ID 5a3f082985eac8305334389600bf64ca7413816ee5bc16b3707ba8cc9d11a59c for node slave1-0007qgsnyzliy from image: jenkins/ssh-slave
```

So basically, containers are provisioned but nothing is executed inside them.

**Comment by pjdarton** [comment]
If the plugin is successfully creating containers _and_ you're getting slave nodes appearing in Jenkins for those containers _but_ those containers aren't coming "online" then there's something wrong with the way you've configured your template.
I suggest that you check the log on the slave node's web-page to see why it's not coming online - If you're using SSH, 9 times out of 10 it's because you didn't set the "host verification strategy" to be the "non-verifying strategy", causing the SSH connection to be refused.

FYI the SSH plugin (now) supports extra security to allow Jenkins to have prior knowledge of all the host keys of all the slave nodes it'll connect to and to insist that anything it connects to is "already known".  Docker slaves are dynamically created and, therefore, typically have their own unique (dynamically-generated) host key which Jenkins has no prior knowledge of.
i.e. While this SSH-connection feature is useful when dealing with a small number of static slaves, it's not much use when slaves are dynamically created.

**Comment by asadpiz** [comment]
Hey thanks for such a quick response, I'm not using the host machine as slave (because it's not gonna run any jenkins agents). My setup is intended to use docker containers as slaves, which as you mentioned are dynamic. So, I'm not sure how to disable the verification strategy for containers. 

So, is it necessary to add the host machine as a node to Jenkins, for this plugin to work? Because the container is supposed to be the ["official" jenkins slave container](https://github.com/jenkinsci/docker-ssh-slave)

**Comment by pjdarton** [comment]
No, this has nothing to do with using the host machine as a slave; this has nothing to do with adding the host machine as a node.
This is all about how you configured your docker template.

Your logs indicate that you're using `jenkins/ssh-slave` so, in the template for that (Manage -> Configure System, scroll down to "Clouds", expand "Docker Templates", find the template you're using), you should have pretty much everything as the default, except:
- Remote File System Root = /home/jenkins/agent
- Connect method = Connect with SSH
- SSH key = Inject SSH key
- User = jenkins

If you're using SSH key setting of "User configured SSH credentials" then you'll need to specify some SSH Credentials that are a valid login for that container's SSH daemon _and_ specify the Host Key Verification Strategy (which should usually be set to Non verifying Verification Strategy).

**Comment by asadpiz** [comment]
So, my settings are pretty much the same except for the: **Remote File System Root = /home/jenkins/agent**
As there's no "agent"  directory in [ssh-slave container](https://github.com/jenkinsci/docker-ssh-slave/blob/master/Dockerfile), I've set it up to be just **/home/jenkins**.

PS: I tried to modify the ssh-slave container to have the /home/jenkins/agent directory but I still get the same problem of Container being up and running but the slave isn't becoming online in Jenkins UI for some reason. 
I'm gonna try with another docker template to see if it works.

Here are my settings: https://imgur.com/a/nPFBiKb

**Comment by asadpiz** [comment]
Just to add System Info:
Jenkins Master: Virtual Machine, RHEL 8, No Docker installed
Jenkins Slave: Virtual Machine, Amazon Linux, Docker Installed, docker port accessible by master

**Comment by pjdarton** [comment]
Last time I looked, `/home/jenkins/agent` was a docker VOL in the Dockerfile.  FYI I found that out "the hard way" because I was trying to be clever and pre-populate various Jenkins tools in a custom image and found that, no matter what I copied into `/home/jenkins/agent/tools`, nothing persisted :grin:
FYI the settings I suggested are also the settings that the [ssh-connection unit test code](https://github.com/jenkinsci/docker-plugin/blob/docker-plugin-1.1.6/src/test/java/io/jenkins/docker/connector/DockerComputerSSHConnectorTest.java) uses, so I'd hope they're correct.

You could also try docker image `jenkins/jnlp-slave` with the JNLP connection method (again, user = jenkins, remote root = /home/jenkins/agent) or docker image `jenkins/slave` with the direct Attach connection method (again, user = jenkins, remote root = /home/jenkins/agent)

FYI the ssh method requires the Jenkins master machine to be able to do a TCP/IP connection to the SSH port exposed by the container on your docker host.  The jnlp method requires the container to be able to do a TCP/IP connection to the master.  The attach method just relies on Jenkins being able to talk to the docker host itself so it'll work even when your networking setup is weird _but_ it'll put additional load on the docker daemon on the host (as it's acting as a middle-man), which some folks might want to avoid.

**Comment by pjdarton** [comment]
Re: "Amazon Linux"
Another "hard won" lesson I learned is that you have to be careful when mixing different Javas.
I found that Oracle & OpenJDK weren't fussy - as long as the major versions were the same, you could have any Oracle running master & any OpenJDK on the slaves, or the other way around _but_ I found that if I was using the IBM Java I had to ensure that I used exactly the same version (including patches etc) everywhere.
I've never used/tried Amazon Linux but I would suggest that you use OpenJDK 8 (or higher) _everywhere_ as that's what Jenkins is built, tested & supported on.

**Comment by asadpiz** [comment]
It worked with docker attach (instead of SSH).
Username = jenkins
remote root = /home/jenkins
As, you correctly pointed out in the first comment, it has to something to do with the way, I've setup the SSH keys. I created the .ssh folder in $JENKINS_HOME and put the public private key-pair their, but i guess that didn't work.

**Comment by pjdarton** [comment]
To debug a SSH docker image, it's best to create a docker container manually _and_ to then create a Jenkins slave node, telling Jenkins it needs to login using SSH to <IP> on <port> (whatever IP:port your container's SSH port is exposed on) so you can manually debug it.
Once you know how to do it manually, then you can tell the docker-plugin to do that automatically.

**Comment by asadpiz** [comment]
So, i was able to ssh (from master) into the standalone container. Still can't figure out why it can't do so via the plugin, so as per my understanding, the **SSH Inject method**, will create a new key-pair and pass the public key to the container. It will then use the private key generated in the previous step, to ssh into this container?

Is this correct? or do I need to somehow, manually create the key-pair and configure it somewhere in Jenkins.

**Comment by pjdarton** [comment]
Yup, that's the way the SSH-Inject method works.  No further steps required.

However, you will need to ensure that the filesystem location into which the plugin puts the key is the same location that the SSH service will be looking for one, and that the ownership etc is correct etc, otherwise the SSH service won't let Jenkins log in.

**Comment by asadpiz** [comment]
So, I was able to launch the container by using user provided credentials (instead of inject SSH).

I discovered that when "inject SSH method" is given, plugin doesn't pass the key to the container correctly (is there a way to see the docker command executed by the plugin?).

The jenkins/ssh-slave [docker file](https://github.com/jenkinsci/docker-ssh-slave/blob/master/Dockerfile), uses a "[setup-sshd](https://github.com/jenkinsci/docker-ssh-slave/blob/master/setup-sshd#L39)" bash script to parse the public key. 

1. When inject SSH is used: docker logs:
```
 + [[ '' == ssh-* ]]
+ [[ 8 -gt 0 ]]
+ [[ /usr/sbin/sshd == ssh-* ]]
+ exec /usr/sbin/sshd -D -p 22 -o AuthorizedKeysCommand=/root/authorized_key -o AuthorizedKeysCommandUser=root
```
and the auhorized_key file is created under /root, instead of /home/jenkins and is populated with :

> #!/bin/sh
> [ "$1" = "jenkins" ] && echo 'ssh-rsa AAAAB3dfdf..<rest of public key>'|| :root@5

the part before ssh-rsa, is problematic.

Apparently the command from docker-plugin doesn't satisfy the condition ([setup-sshd#L39](https://github.com/jenkinsci/docker-ssh-slave/blob/master/setup-sshd#L39)): 

`if [[ $JENKINS_SLAVE_SSH_PUBKEY == ssh-* ]];`

this causes the issues.
